### PR TITLE
GLM: add kunlunxin R300x1x8 config

### PR DESCRIPTION
**Pull request by dynamicheart**

### CODE REVIEW

**Comment by dynamicheart** [comment]
rank0 log:
```
[PerfLog] {"event": "STEP_END", "value": {"loss": 0.2211799919605255, "eval_accuracy": 0.7768999934196472, "epoch": 0, "end_training": false, "global_steps": 2399, "num_trained_samples": 76768, "learning_rate": 5.469801751959428e-06, "seq/s": 3.9555062554611267}, "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 110, "time_ms": 1679067812797, "rank": 0}}
[PerfLog] {"event": "STEP_END", "value": {"loss": 0.6775041818618774, "eval_accuracy": 0.8003999590873718, "epoch": 0, "end_training": true, "global_steps": 2400, "num_trained_samples": 76800, "learning_rate": 5.467496542185339e-06, "seq/s": 3.946610529871572}, "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 110, "time_ms": 1679068444192, "rank": 0}}
[PerfLog] {"event": "EVALUATE", "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 110, "time_ms": 1679068444192, "rank": 0}}
[PerfLog] {"event": "EPOCH_END", "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 110, "time_ms": 1679068444468, "rank": 0}}
[PerfLog] {"event": "TRAIN_END", "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 124, "time_ms": 1679068444469, "rank": 0}}
[PerfLog] {"event": "FINISHED", "value": {"e2e_time": 22902.81707048416, "training_sequences_per_second": 3.4690496628788376, "converged": true, "final_accuracy": 0.8003999590873718, "raw_train_time": 22138.628, "init_time": 762.763}, "metadata": {"file": "/workspace/flagperf/training/benchmarks/glm/pytorch/run_pretraining.py", "lineno": 143, "time_ms": 1679068444501, "rank": 0}}
```

kunlunxin_monitor log:
```
2023-03-17-23:53:39
46C 118W 22446MiB 32768MiB 100%
40C 119W 9346MiB 32768MiB 100%
42C 120W 26629MiB 32768MiB 100%
46C 126W 17439MiB 32768MiB 99%
46C  95W 30813MiB 32768MiB 100%
42C 106W 29784MiB 32768MiB 100%
42C 134W 4542MiB 32768MiB 60%
45C  76W 30881MiB 32768MiB 100%

2023-03-17-23:53:44
44C 119W 29784MiB 32768MiB 100%
40C 107W 29784MiB 32768MiB 100%
43C 116W 30744MiB 32768MiB 100%
48C 124W 31400MiB 32768MiB 100%
49C 121W 31637MiB 32768MiB 100%
42C 120W 4542MiB 32768MiB 0%
42C 120W 30539MiB 32768MiB 100%
48C 131W 31400MiB 32768MiB 100%
```

[rank0.out.log](https://github.com/FlagOpen/FlagPerf/files/11007263/rank0.out.log)
[kunlunxin_monitor.log](https://github.com/FlagOpen/FlagPerf/files/11007264/kunlunxin_monitor.log)

File: training/kunlunxin/docker_image/pytorch/Dockerfile.source [hunk 1]
```diff
@@ -0,0 +1,49 @@
+FROM ubuntu:18.04
```
Comment (yuzhou03) [inline training/kunlunxin/docker_image/pytorch/Dockerfile.source:1]: source code to build image: dynamicheart/pytorch1.12.1-cpu-ubuntu18.04:v0.04

File: training/kunlunxin/glm-pytorch/extern/layers/transformer.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        return (output, mem_layers)
+
+    def update_mems(self, hiddens, mems, return_memory=False):
+        memory_length = mems[0].size(1) if mems else 0
+        query_length = hiddens[0].size(1)
+        new_memory_length = memory_length + query_length
+        if not return_memory:
+            new_memory_length = min(self.max_memory_length, new_memory_length)
+        new_mems = []
+        # with torch.no_grad():
```
Comment (yuzhou03) [inline training/kunlunxin/glm-pytorch/extern/layers/transformer.py:258]: commented code? if yes, should be removed.

File: training/kunlunxin/glm-pytorch/extern/layers/transformer_block.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                          output_layer_init_method=output_layer_init_method)
+
+    def forward(self,
+                hidden_states,
+                ltor_mask,
+                position_embeddings=None,
+                r_w_bias=None,
+                r_r_bias=None,
+                mem=None):
+        # hidden_states: [b, s, h]
```
Comment (yuzhou03) [inline training/kunlunxin/glm-pytorch/extern/layers/transformer_block.py:85]: Comment for method parameters should be a block comment. like
image

File: training/kunlunxin/glm-pytorch/extern/layers/transformer.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+                attention_scale=attention_scale)
+
+        # Transformer layers.
+        self.layers = torch.nn.ModuleList(
+            [get_layer() for _ in range(num_layers)])
+
+        # Final layer norm before output.
+        self.final_layernorm = LayerNorm(hidden_size, eps=layernorm_epsilon)
+
+        # if deepspeed.checkpointing.is_configured():
```
Comment (yuzhou03) [inline training/kunlunxin/glm-pytorch/extern/layers/transformer.py:134]: commented code

File: training/kunlunxin/glm-pytorch/extern/converter.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+from driver import dist_pytorch
+from .layers.transformer import GLMTransformer
+
+
+def convert_model(model, config):
+    if dist_pytorch.get_rank() == 0:
+        print("convert model", flush=True)
+    model = model.cpu()
+    state_dict = model.state_dict()
+    # the initilization of GLMTransformer will change the rng state
```
Comment (yuzhou03) [inline training/kunlunxin/glm-pytorch/extern/converter.py:12]: typo: initialization

File: training/kunlunxin/glm-pytorch/extern/layers/transformer.py (hunk 2 above)
Comment (dynamicheart) [reply training/kunlunxin/glm-pytorch/extern/layers/transformer.py:258]: This file is mainly copied from the `benchmarks` directory and so is this line of commented code. As discussed, these code will be kept until we figure out a new method to manage the model modifications.

File: training/kunlunxin/glm-pytorch/extern/layers/transformer_block.py (hunk 3 above)
Comment (dynamicheart) [reply training/kunlunxin/glm-pytorch/extern/layers/transformer_block.py:85]: This file is mainly copied from the `benchmarks` directory and so is this block comment. As discussed, these code swill be kept until we figure out a new method to manage the model modifications.

File: training/kunlunxin/glm-pytorch/extern/layers/transformer.py (hunk 4 above)
Comment (dynamicheart) [reply training/kunlunxin/glm-pytorch/extern/layers/transformer.py:134]: This file is mainly copied from the `benchmarks` directory and so is this line of commented code. As discussed, these codes will be kept until we figure out a new method to manage the model modifications.

File: training/kunlunxin/glm-pytorch/extern/converter.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    with torch.random.fork_rng():
```
[the remark this answers is not in the document]
Comment (dynamicheart) [reply training/kunlunxin/glm-pytorch/extern/converter.py:13]: will remove this code

File: training/kunlunxin/glm-pytorch/extern/layers/transformer.py (hunk 4 above)
Comment (yuzhou03) [reply training/kunlunxin/glm-pytorch/extern/layers/transformer.py:134]: OK

File: training/kunlunxin/docker_image/pytorch/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,2 @@
+from dynamicheart/pytorch1.12.1-cpu-ubuntu18.04:v0.04
```
[the remark this answers is not in the document]
Comment (dynamicheart) [reply training/kunlunxin/docker_image/pytorch/Dockerfile:1]: will add a `TODO` comment

File: training/kunlunxin/docker_image/pytorch/Dockerfile.source [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
```
[the remark this answers is not in the document]
Comment (dynamicheart) [reply training/kunlunxin/docker_image/pytorch/Dockerfile.source:2]: got it

**Review by upvenly — APPROVED** [review]

**Pull request merged**